src/kernels/ipywidgets-message-coordination/localIPyWidgetScriptManager.node.ts

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import * as path from '../../platform/vscode-path/path';
+import { Uri } from 'vscode';
+import { IFileSystemNode } from '../../platform/common/platform/types.node';
+import { IExtensionContext } from '../../platform/common/types';
+import { StopWatch } from '../../platform/common/utils/stopWatch';
+import { sendTelemetryEvent, Telemetry } from '../../telemetry';
+import { getTelemetrySafeHashedString } from '../../telemetry/helpers';
+import { IKernel } from '../types';
+import { BaseIPyWidgetScriptManager } from './baseIPyWidgetScriptManager';
+import { IIPyWidgetScriptManager, INbExtensionsPathProvider } from './types';
+
+type KernelConnectionId = string;
+export class LocalIPyWidgetScriptManager extends BaseIPyWidgetScriptManager implements IIPyWidgetScriptManager {
+    /**
+     * Copy once per session of VS Code or until user restarts the kernel.
+     */
+    private sourceNbExtensionsPath?: Uri;
+    private overwriteExistingFiles = true;
+    static nbExtensionsCopiedKernelConnectionList = new Set<KernelConnectionId>();
+    constructor(
+        kernel: IKernel,
+        private readonly fs: IFileSystemNode,
+        private readonly nbExtensionsPathProvider: INbExtensionsPathProvider,
+        private readonly context: IExtensionContext
+    ) {
+        super(kernel);
+        // When re-loading VS Code, always overwrite the files.
+        this.overwriteExistingFiles = !LocalIPyWidgetScriptManager.nbExtensionsCopiedKernelConnectionList.has(
+            kernel.kernelConnectionMetadata.id
+        );
+    }
+    public getBaseUrl() {
+        return this.getNbExtensionsParentPath();
+    }
+    protected async getNbExtensionsParentPath(): Promise<Uri | undefined> {
+        let overwrite = this.overwriteExistingFiles;
+
+        try {
+            const stopWatch = new StopWatch();
+            this.sourceNbExtensionsPath = await this.nbExtensionsPathProvider.getNbExtensionsParentPath(this.kernel);
+            if (!this.sourceNbExtensionsPath) {
+                return;
+            }
+            const kernelHash = getTelemetrySafeHashedString(this.kernel.kernelConnectionMetadata.id);
+            const baseUrl = Uri.joinPath(this.context.extensionUri, 'tmp', 'scripts', kernelHash, 'jupyter');
```

Comment:
Not sure where to put the comment but it might make sense to say why we need to copy the file. It's because we can't load from disk in the webview. We need a toWebviewUri possible location.

Reply:
I'll put the comment at the class level